Cabal-syntax-3.8.1.0: allow mtl-2.3 and transformers-0.6

andreasabel opened the pull request:
The released Cabal-syntax 3.8.1.0 is already compatible with mtl-2.3.  I am intending to make the respective revision on Hackage (details see this PR).  This PR updates the 3.8 branch accordingly.

Lifted from: https://github.com/haskell/hackage-security/pull/289#issuecomment-1367200577

@Mikolaj wrote:
> if it's useful to anybody, I can revise even omitting the comment (but we should probably amend 3.8 branch, too, which has old bounds, so that somebody compiling from 3.8 branch source doesn't have worse bounds than from Hackage).

(I'd think the said comment can remain.)

Mikolaj reviewed (approved):
This is fine. If CI fails (it may not be up to date with the recent workarounds we needed for master branch), I will force-merge.

andreasabel commented:
I suggest to **SQUASH** the 2 commits. 

Strange, I cannot push this `Cabal-syntax.cabal` to Hackage:
```
$ hackage-cli push-cabal Cabal-syntax.cabal 
...
Errors:

Cannot change the &#39;if&#39; condition expressions
```

So maybe the `3.8` branch isn't in sync with the Hackage release of `Cabal-syntax-3.8.1.0`...

Anyway, one can make a manual revision.  
However, then I am not so sure whether putting `x-revision: 2` in here makes sense still.

Mikolaj commented:
You are right, it's not in sync. I neglected reflecting the revision on the branch: https://hackage.haskell.org/package/Cabal-syntax-3.8.1.0/revisions/. Naughty maintainer, no cookie.

andreasabel commented:
> You are right, it's not in sync. I neglected reflecting the revision on the branch: [hackage.haskell.org/package/Cabal-syntax-3.8.1.0/revisions](https://hackage.haskell.org/package/Cabal-syntax-3.8.1.0/revisions/). 

Well, I include revision 1 in this PR, but this does not explain the error I am getting from hackage.  There is another difference:
```diff
54a55,56
> 
>   if impl(ghc >= 8.0)
57c59
<   if impl(ghc <8.8)
---
>   if impl(ghc >= 8.0) && impl(ghc < 8.8)
```
The branch 3.8 has: https://github.com/haskell/cabal/blob/399cf6432df746b0b0d8e33a8b41736120552380/Cabal-syntax/Cabal-syntax.cabal?rgh-link-date=2022-12-29T09%3A56%3A29Z#L55-L59
Hackage has:
```cabal
  ghc-options: -Wcompat -Wnoncanonical-monad-instances

  if impl(ghc <8.8)
    ghc-options: -Wnoncanonical-monadfail-instances
```
Hackage is optimized because `ghc >= 8.0` is essentially implied by `base >= 4.9`.

It makes sense to conform the 3.8 branch to the Hackage variant.  I'll push another commit to that extend.

andreasabel commented:
Ok, after synchronizing the conditionals a revision is possible.  Dry-run produces this answer from Hackage:

<ul>
  
 * Changed the library component&#39;s library dependency on &#39;mtl&#39;
                          from `>=2.1 && <2.3`
                          to `>=2.1 && <2.4`

 * Changed the library component&#39;s library dependency on &#39;transformers&#39;
                          from `>=0.3 && <0.4 || >=0.4.1.0 && <0.6`
                          to `>=0.3 && <0.4 || >=0.4.1.0 && <0.7`

</ul>

Mikolaj reviewed (approved):
Still LGTM.

Mikolaj commented:
Doctest CI supposedly still running, but I think it's stuck and everything is finished. Shall I squash+merge?

andreasabel commented:
> Shall I squash+merge?

Yes, please!  Thanks.

Here is the Hackage revision: https://hackage.haskell.org/package/Cabal-syntax-3.8.1.0/revisions/

Mikolaj commented:
> Here is the Hackage revision: https://hackage.haskell.org/package/Cabal-syntax-3.8.1.0/revisions/

Splendid. Thank you.
